README.md
# ADC-Adjust
Calibrate ADC for various MCUs.  Develop with PlatformIO

   AdcAdjust is a tool for evaluating ADC performance of various
   MCUs. Measurement noise is reduced by averaging. Two methods are
   provided for fixing measurement offset errors.  One, It is best to have
   a reliable DVM to measure the input test voltage. Two, without
   a DVM, a LM35 sensor and a thermometer can provide a known input
   voltage. Then a sofware parameter can fix the error.

   PlatformIO enables various MCUs to share the same code package.

   Source code is formatted with Doxygen markup.  Doxygen is a very
   good tool for tracing code and linking embedded images.


Changelog.txt
/** \page changes_ AdcAdjust Changelog

\par Doxygen Procedures
- cd ~/pio/AdcAdjust (for example)
- cmd: doxygen Doxyfile or doxywizard Doxyfile
- see \ref notes_install_
- Get Doxygen to generate html document set with schematics

== changes_ */


//--------------------------------------------------------------------
/**  @mainpage 

\anchor top_

  \par AdcAdjust Introduction

   AdcAdjust is a tool for evaluating ADC performance of various
   MCUs. Measurement noise is reduced by averaging. Two methods are
   provided for fixing measurement offset errors.  One, It is best to have
   a reliable DVM to measure the input test voltage. Two, without
   a DVM, a LM35 sensor and a thermometer can provide a known input
   voltage. Then a sofware parameter can fix the error.

   PlatformIO enables various MCUs to share the same code package.

   Source code is formatted with Doxygen markup.  Doxygen is a very
   good tool for tracing code.

\par Technology

 -  Arduino UNO variants, ESP32 variants
 -  Optional LM35DZ temperature sensor for test
 -  PlatformIO, Arduino and ESP-IDF MCU libraries
 -  Doxygen
 
.. | Information
:--------- | :----------------------------
Project    | AdcAdjust
License    | Open Source
Version    | (release candidate 01)
Author     | Robert M. McIsaac; Software Designer
Date       | 2022.mar.24
<br>

\note
  - See \ref notes_install_ as starting point for program code
  - See \ref changes_ for daily changelog

end Mainpage 
*/ 

/** \page notes_install_ Installation

<br>

\verbatim

   The AdcAdjust code can be copied into any folder. In addition to
   'platformio.ini', a convenient 'Makefile' is provided which selects
   the MCU and remembers pio commands to initialize, compile, and
   upload the AdcAdjust firmware.
   
\endverbatim <br>
 
\par MCU Boards

 - USE_MCU=1; metro (mini328) enhanced UNO
 - USE_MCU=2; nano33ble sense
 - USE_MCU=3; featheresp32
 - USE_MCU=4; esp32doit-devkit-v1-D 
 - USE_MCU=5; esp32dev

\verbatim
  Edit the Makefile to select a board. Only one at a time is valid.
  Example: "BRD := esp32dev"

\endverbatim
<br>
 
\par Arduino MCU Board ADC Adjustment
\verbatim
   
  Arduino Analog to digital conversion (ADC), is done in 3 steps:-

  - A voltage is applied to a GPIO pin which is assigned to ADC input.
  - A program reads the count of ADC steps
  - The program calculates the voltage using a linear function

  Voltage = ADC_VALUE * REFERENCE / MAX_ADC_STEPS

  There is some internet debate about MAX_ADC_STEPS for a 10-bit
  ADC. Some say it should be 1023 and others say 1024 is
  best. However, AdcAdjust uses 1024 because there are 1024 states from
  no-bits-set (zero) to all-bits-set (1023). This debate is probably
  moot because of intrinsic error factors. See AN5250.

  AdcAdjust replaces REFERENCE with OPERATING_VOLTAGE (.opV). This
  is because the reference is only a nominal value (1.1vdc) that
  varies between MCU boards. If the ADC printed voltage is higher than
  the input, then reduce the REFERENCE parameter by a few
  millivolts. For example, the REFERENCE for the MetroMini328 needed
  to be (1.07vdc) in routine "metro_init".

  \endverbatim \ref metro_init()<br>
  https://www.nxp.com/docs/en/application-note/AN5250.pdf <br><br>

\par ESP32 MCU Board ADC Adjustment
\verbatim

   AdcAdjust uses the Espressive Software Development Kit (SDK)
   functions for ESP32 ADC. The application needs to call
   "adc_sdk_sample()" which returns millivolts. If the ADC can be
   plotted by "y=mx+b" then it seems the "b" term has an offset, which
   should be zero. See "doit_read()" which subtracts 10 millivolts as an
   example.

\endverbatim \ref doit_read()
<br><br>

\par MCU Board ADC Test
\verbatim

   A KiCad schematic of the test voltage generator is provided. There
   are two outputs:-- (0.7vdc) and (0.2vdc) connected one at a time to
   the ADC GPIO pin.  These are arbitrary values.  Two points are
   recommended because a check of only one point could work
   accidently.

\endverbatim
\image html adc_testpoints.jpg
<br>

\verbatim

   A reliable DVM is needed for exact measure of the test voltages. If
   no DVM is available, the divider circuit is not needed. Optionally,
   a LM35DZ can provide a voltage representing temperature. The
   program parameters can be adjusted so that the printed temperature
   (volts * 100) matches a nearby trusted thermometer.

   The adc_print() prints four CSV columns:-- LineNumber, ADC value,
   average ADC value, and average voltage. These values can be logged
   in a CSV file and viewed in a spreadsheet. The following image is a
   plot of ESP32 ADC values in columns B and C. Note that ESP32 has
   periodic spikes which might be caused by WiFi activity. The Roll
   Averaging compresses the ADC noise and spikes which can be ignored
   by an application. The cleaned ADC voltage varies only 2
   millivolts.
   
 \endverbatim 
 \ref adc_print in file adc_mv.cpp <br>

  \image html ESP32_ADC_test.jpg <br>

  \image html ESP32_ADC_varies_ok.jpg
 
*/


src/adc_sdk.cpp
// -*- C++ -*-
//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
/**@file  adc_sdk.cpp 
 * @brief ESP32 analog to digital (ADC) using Espressif SDK
 * 
 */ 

#include "adc_sdk.h" 
#define CHOICES (USE_MCU==3 || USE_MCU==4 || USE_MCU==5)
#if CHOICES > 0
// the macros are not defined for Doxygen; so nothing extracted

#include "driver/adc.h"
#include "esp_adc_cal.h"
#include "adc_mv.h" 
#include "run_average.h"

enum { DEFAULT_VREF = 1100 };
static esp_adc_cal_characteristics_t *adc_chars;
static u32t adc_voltage=0, adc_reading = 0;

static void check_efuse(void);
static void print_char_val_type(esp_adc_cal_value_t val_type);
int   adc_sdk_init(void);
u32t  adc_sdk_sample(void);

//---------------------------------------------------------------
/// \brief Check if TP is burned into eFuse (it is not)
static void check_efuse(void){
   if (esp_adc_cal_check_efuse(ESP_ADC_CAL_VAL_EFUSE_TP) == ESP_OK) {
      printf("eFuse Two Point: Supported\n");
   } else {
      printf("eFuse Two Point: NOT supported\n");
   }
   //Check Vref is burned into eFuse
   if (esp_adc_cal_check_efuse(ESP_ADC_CAL_VAL_EFUSE_VREF) == ESP_OK) {
      printf("eFuse Vref: Supported\n");
   } else {
      printf("eFuse Vref: NOT supported\n");
   }
}

//---------------------------------------------------------------
/// test if calibration in eFuse (it is not)
static void print_char_val_type(esp_adc_cal_value_t val_type){
   if (val_type == ESP_ADC_CAL_VAL_EFUSE_TP) {
      printf("Characterized using Two Point Value\n");
   } else if (val_type == ESP_ADC_CAL_VAL_EFUSE_VREF) {
      printf("Characterized using eFuse Vref\n");
   } else {
      printf("Characterized using Default Vref\n");
   }
}

//---------------------------------------------------------------
/// \brief Init the SDK ADC parameters
int adc_sdk_init(void){
   
   static const adc_bits_width_t width = ADC_WIDTH_BIT_12;
   static const adc_atten_t      atten = ADC_ATTEN_DB_0;
   static const adc_unit_t       unit  = ADC_UNIT_1;
 
   check_efuse();

   adc1_config_width(width);
   adc1_config_channel_atten((adc1_channel_t)ADC_CHANNEL_6, atten);
   void *p = calloc(1, sizeof(esp_adc_cal_characteristics_t));
   adc_chars = (esp_adc_cal_characteristics_t *)p;
   esp_adc_cal_value_t val_type;
   val_type = esp_adc_cal_characterize(
				       unit,
				       atten,
				       width,
				       DEFAULT_VREF,
				       adc_chars);
   print_char_val_type(val_type);
   return 0;
}

//---------------------------------------------------------------
/// \brief sample ADC1 ch6 (IO34)
/// \return adc_voltage ..millivolts; requires RunAverage to reduce noise
u32t adc_sdk_sample(void){
   //check_efuse();
   adc_reading = adc1_get_raw((adc1_channel_t)ADC_CHANNEL_6);
   adc_voltage = esp_adc_cal_raw_to_voltage(adc_reading, adc_chars);
   return adc_voltage;
}
#endif // CHOICES


//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 

#if CHOICES==0

/// dummy function when ESP32 not used
int   adc_sdk_init(void){
   return 0;
}
/// dummy function when ESP32 not used
u32t  adc_sdk_sample(void){
   return 0;
}
#endif



src/run_average.h
// -*- C++ -*-
//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**@file  run_average.h
 * @brief to reduce measurement noise
 *
 * 2022.mar.21
 * @author RMM
 */
#if !defined RunAverage_h
#    define RunAverage_h
#include "Main.h"

//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/// \brief smooth data with a running average
class RunAverage {

private:
   int NR;
   int idx;                     // array index
   float *pBuff;		// the buff from the analog input
   float total;			// the running total
   bool seeded;                 // read NR times to prepare array

public:
   RunAverage(int size);
   ~RunAverage(void);

private:
   /// @brief update average, return latest value
   float add_buff( float num );
   
public:
   /// @brief add number to average, return latest value
   float new_num( float num );

};

#endif //    define RunAverage_h


src/run_average.cpp
// -*- C++ -*-
//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**@file  run_average.cpp
 * @brief to reduce measurement noise
 *
 * 2022.mar.21
 * @author RMM
 */
#include "run_average.h"

//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/// \brief smooth data with a running average
RunAverage::RunAverage(int size){
   NR = size;
   pBuff = new float[NR];
   idx = 0;
   total = 0;
   seeded = false;
}

RunAverage::~RunAverage(void){
   delete [] pBuff;
}

/// @brief update average, return latest value
float RunAverage::add_buff( float num ){
   total = total - pBuff[idx];
   pBuff[idx] = num;
   total = total + num;
   if(++idx >= NR) {
      idx= 0;
   }
   float avg = total*1.0 / NR;
   return avg;
}
   
/// add the latest data to the averaging array
float RunAverage::new_num( float num ) {
   if( !seeded ){
      idx = 0;
      total = 0;
      for( int i = 0; i < NR; i++ ){
	 pBuff[i] = num;
	 total += num;
      }
      seeded = true;
   }
   float avg = add_buff(num);
   return avg;
}

// run_average.cpp



src/adc_mv.cpp
// -*- C++ -*-
//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
/**@file  adc_mv.cpp 
 * @brief to measure incoming voltage of ADC GPIO 
 * 
 */ 
#include "adc_mv.h" 
#include "run_average.h"
#include "adc_sdk.h"

typedef struct {
   unsigned usteps;	// ADC steps from analogRead() or system_read()
   unsigned max_steps;  // say 10-bits = 1024;  12bits = 4096
   float opV;		// ADC operating voltage or reference
   float volts;		// measured volts
   float offset;	// another adjustable
}ADC_T;			// analog to digital conversion
   
static RunAverage Avg1(32); // average voltage
static RunAverage Avg2(32); // average ADC steps
static ADC_T adc;
static void adc_print(int skip_if_zero);

static struct {
   unsigned numReads;
   int pin_blink;
   int pin_adc;
}my;

//-----------------------------------------------------------------
/// prepare to test ADC of the Adafruit MetroMini328
void metro_init(void){

#if USE_MCU==1
   my.numReads = 0;
   my.pin_blink = 13;
   my.pin_adc = A4;
   pinMode(my.pin_blink, OUTPUT);
   pinMode(my.pin_adc, INPUT);
   analogReference(INTERNAL); // 1.1vdc nominal (approximately)
   adc.opV = 1.07;		// adjust to improve result
   adc.max_steps = 1024;	// 0 to 1023
   adc.offset = 0.00;		// another adjustable
   return;
#endif // USE_MCU==1
}

/// \brief convert ADC steps to voltage
void metro_read(){
#if USE_MCU==1
   adc.usteps = analogRead(my.pin_adc);
   if( adc.max_steps < 1 ) return;
   float fv =  (adc.opV / adc.max_steps) * adc.usteps + adc.offset;
   adc.volts = Avg1.new_num(fv);
   adc_print(1);
   return;
#endif // USE_MCU==1
}

//-----------------------------------------------------------------
/// \brief prepare to use the NANO 33 BLE sense nRF52840
void nano33_init(void){

#if USE_MCU==2
   my.numReads = 0;
   my.pin_blink = 13;
   my.pin_adc = A1;
   
   pinMode(my.pin_blink, OUTPUT);
   pinMode(my.pin_adc, INPUT);
   analogReadResolution(10);	// 0 to 1023
   analogReference(AR_INTERNAL1V2);
   adc.opV = 1.20;		// adjust to improve result
   adc.max_steps = 1024;	// for 10 bits
   adc.offset = 0.00;		// another adjustable
   return;
#endif // USE_MCU==2
}

/// \brief convert ADC steps to voltage
void nano33_read(void){
#if USE_MCU==2			// NANO 33 BLE nRF52840
   int flag = my.numReads++ & 1;
   digitalWrite(my.pin_blink, flag);
   adc.usteps = analogRead(my.pin_adc);
   if( adc.max_steps < 1 ) return;
   float fv =  (adc.opV / adc.max_steps) * adc.usteps + adc.offset;
   adc.volts = Avg1.new_num(fv);
   adc_print(1);
   return;
#endif // USE_MCU==2
}

//-----------------------------------------------------------------
/// \brief prepare to use the Adafruit Feather ESP32
void feather32_init(void){
#if USE_MCU==3
   my.numReads = 0;
   my.pin_blink = 13;
   my.pin_adc = 34;
   pinMode(my.pin_blink, OUTPUT);
   pinMode(my.pin_adc, INPUT);
   //adc.offset = 0.00;
   adc_sdk_init();
   return;
#endif // USE_MCU==3
}

/// \brief convert ADC steps to voltage
void feather32_read(void){
#if USE_MCU==3
   int flag = my.numReads++ & 1;
   digitalWrite(my.pin_blink, flag);
   adc.usteps =  adc_sdk_sample();  // get millivolts from SDK
   float fv = adc.usteps - 10;  // adjust millivolts
   adc.volts = Avg1.new_num(fv/1000.0);
   adc_print(1);
   return;
#endif // USE_MCU==3
}

//-----------------------------------------------------------------
/// \brief prepare to use the Doit devel board
void doit_init(void){

#if USE_MCU==4			// esp32doit-devkit-v1
   my.numReads = 0;
   my.pin_blink = 02;
   my.pin_adc = 34;
   pinMode(my.pin_blink, OUTPUT);
   pinMode(my.pin_adc, INPUT);
   // adc.opV = 1.20;		// adjust to improve result
   // adc.max_steps = 4096;	// 
   // adc.offset = 0.00;
   adc_sdk_init();
   return;
#endif // USE_MCU==4
}

/// \brief convert ADC steps to voltage
void doit_read(void){
#if USE_MCU==4
   int flag = my.numReads++ & 1;
   digitalWrite(my.pin_blink, flag);
   adc.usteps =  adc_sdk_sample();  // get millivolts from SDK
   float fv = adc.usteps - 10;  // adjust millivolts
   adc.volts = Avg1.new_num(fv/1000.0);
   adc_print(1);
   return;
#endif // USE_MCU==4
}

//-----------------------------------------------------------------
/// \brief prepare to use the Espressif ESP32
void esp32dev_init(void){

#if USE_MCU==5			// ESP32 DevKit V4
   my.numReads = 0;
   my.pin_blink = 02;
   my.pin_adc = 34;
   pinMode(my.pin_blink, OUTPUT);
   pinMode(my.pin_adc, INPUT);
   //  adc.offset = 0.00;
   adc_sdk_init();
   return;
#endif // USE_MCU==5
}

/// \brief convert ADC steps to voltage
void esp32dev_read(void){
#if USE_MCU==5
   int flag = my.numReads++ & 1;
   digitalWrite(my.pin_blink, flag);
   adc.usteps =  adc_sdk_sample();  // get millivolts from SDK
   float fv = adc.usteps - 2;  // adjust millivolts
   adc.volts = Avg1.new_num(fv/1000.0);
   adc_print(1);
   return;
#endif // USE_MCU==5
}

//-----------------------------------------------------------------
/// \brief prepare to do ADC measurement
float adc_mv_init(void){
   memset(&adc,0,sizeof(adc));
   // code is wrapped in macros so only one these does anything
   metro_init();
   nano33_init();
   feather32_init();
   doit_init();
   esp32dev_init();
   return 0;
}

//------------------------------------------------------------------
/// \brief get analog voltage measurement from GPIO pin
float adc_mv_read(void){
   // code is wrapped in macros so only one these does anything
   metro_read();
   nano33_read();
   feather32_read();
   doit_read();
   esp32dev_read();
   adc_print(0);		// prevent not-used warning
   return adc.volts;
}

//----------- adc_print() ------------------------------------------ 
/// print result of ADC voltage measurement
static void adc_print(int option){
   u32t avgSteps = Avg2.new_num(adc.usteps);
   if(option==0) return;
   if(1){ // This is needed for testing. Else, edit "if(0)"
      static unsigned numLines;
      numLines++;
      String gap = ", ";
      Serial.println( String(numLines) + gap + // line count
		      String(adc.usteps) + gap + // ADC steps
		      String(avgSteps) + gap +	 // average steps
		      String(adc.volts,3)	 // final voltage
		      );
   }
   
}

// end adc_mv.cpp
